File: 0069-sqrtx/0069-sqrtx.py
class Solution:
    def mySqrt(self, x: int) -> int:
        lo, hi = 1, x

        while lo <= hi:
            mid = (lo + hi) // 2
            if mid * mid == x: return mid
            if mid * mid < x: lo = mid + 1
            else: hi = mid - 1

        return lo - 1

File: 0069-sqrtx/0069-sqrtx.java
class Solution {
    public int mySqrt(int x) {
        int lo = 1, hi = x;

        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            long square = (long)mid * mid;
            if (square == x) {
                return mid;
            }
            if (square < x) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        return lo - 1;
    }
}